src/backend/optimizer/path/allpaths.c

```diff
@@ -1700,7 +1700,7 @@ set_subquery_pathlist(PlannerInfo *root, RelOptInfo *rel,
 		 * cannot pass params across motion.
 		 */
 		if ((!bms_is_empty(required_outer)) &&
-			is_query_contain_limit_groupby(subquery))
+			is_query_contain_limit_groupby(subquery, rel->lateral_vars, 1))
```

Comment:
1.
>'is_query_contain_limit_groupby' checks that the sub-query really accesses at least one lateral var;

Do we really need to check this? Give an example when subquery does not contain any lateral vars when `required_outer` is not empty? Maybe cheking for aggregates is enough?

2.
>'is_query_contain_limit_groupby' checks for presence of aggregate functions
at current level.

If the answer to the first question is afirmative, and we must check lateral vars, then why to check aggregates? How certain can we be that only aggregates are to blame in all the troubles with lateral? Are there any cases when we have lateral vars without aggregates?

Reply:
> Maybe cheking for aggregates is enough?

Yes, it looks so. Reworked the patch.

Reply:
https://github.com/arenadata/gpdb/blob/d136ce0cae8945b413f95df826f2a4b4eb3a0bc4/src/backend/optimizer/path/allpaths.c#L3011

How it takes into account nested subqueries? And under term "aggregates"  in "Maybe cheking for aggregates..." I meant the general concept of queries with group by clause. 
```
explain (costs off) select from (values ('')) _out(b)
join lateral (
    select j from (
        select avg(c) as j from t1 
        where t1.b = _out.b group by b
    ) s order by j       
) l on true;

                                        QUERY PLAN                                        
------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)
   ->  Nested Loop
         ->  Values Scan on "*VALUES*"
         ->  Materialize
               ->  Sort
                     Sort Key: s.j
                     ->  Subquery Scan on s
                           ->  GroupAggregate
                                 Group Key: t1.b
                                 ->  Sort
                                       Sort Key: t1.b
                                       ->  Redistribute Motion 3:3  (slice1; segments: 3)
                                             Hash Key: t1.b
                                             ->  GroupAggregate
                                                   Group Key: t1.b
                                                   ->  Seq Scan on t1
                                                         Filter: (b = "*VALUES*".column1)
 Optimizer: Postgres query optimizer
(18 rows)

```

Reply:
When I was checking the latest change, I tried to reproduce the issue with nested subqueries, but was not able to do it (thus I removed recursion into RTEs). I guess the reason I couldn't reproduce the issue, while your query can do it, is that I didn't have 'order' clause in select. So, it looks we need to bring the code:
```
	foreach(lc, parse->rtable)
	{
		RangeTblEntry *rte = (RangeTblEntry *) lfirst(lc);

		if (rte->rtekind == RTE_SUBQUERY &&
			is_query_contain_limit_groupby(rte->subquery))
			return true;
	}
```
back.

Reply:
> And under term "aggregates" in "Maybe cheking for aggregates..." I meant the general concept of queries with group by/limit clause .

Checking for groupby/limit is already there.